Storage Usage: Add benchmark results from v5.10.2 (#214)
crate/cratedb-guide: docs/performance/storage.md

| 2        | 1704074134000        | 1704074399000         | 1               | 1.08          | 1          | "N"                | 249          | 68           | 1            | 7.2         | 1     | 0.5     | 2.44       | 0            | 1                     | 14.64        | 2.5                  | 0           |
```

The taxi dataset takes:
- ~48MiB in Parquet (very optimized for storage)
- ~342MiB in CSV
- ~1.2GiB in JSON
- ~510MiB in PostgreSQL 16.1 (Debian 16.1-1.pgdg120+1)
- ~775MiB in CrateDB 5.9.3 (3 nodes, default settings)
- ~431MiB in CrateDB 5.10.2 (3 nodes, default settings)

We will dive deeper to really understand what is going on.

> > INFO: In version 5.10 storage usage was improved, some users report up to 70% of storage reduction,
> > more reasonably a 30–50% can be expected for most data, the reduction % of reduction will vary
> > depending on use cases. 

Reviewer: FYI: In MyST Markdown, you may want to phrase the [admonition](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html) like this:
```suggestion
:::{note}
In version 5.10 storage usage was improved, some users report up to 70% of storage reduction,
more reasonably a 30–50% can be expected for most data, the reduction % of reduction will vary
depending on use cases. 
:::
```

When a table is created, data is sharded and distributed among nodes. This
means that the memory footprint depends on our replication and sharding strategy.

Let's break down how the `431MiB` in CrateDB and the `510MiB` in PostgreSQL were
obtained. For `PostgreSQL` it was straightforward:

```sql
SELECT pg_size_pretty(pg_total_relation_size('taxi_january'));
```

In CrateDB, when a table is created, sharding and replication have to be taken into account.

> When a table is created with default values, it gets partitioned using this formula: `max(4, num_data_nodes * 2)`.

Reviewer: Maybe elegantly refer to the sharding and scaling guideline docs at this spot?

- https://cratedb.com/docs/guide/performance/sharding.html
- https://cratedb.com/docs/guide/performance/scaling.html#number-of-shards

GROUP BY node ['name']
```

The total storage being used by a table can be calculated as the `average size of 1 shard` * `total_shards`.

Applying this to the `taxi` table that was created beforehand:

```sql
SELECT sum(size / (1024 * 1024)) / count(*) as avg_mib_per_shard,
       sum(size) / (1024 * 1024) as total_mib
FROM sys.shards
WHERE table_name = 'taxi'
```

CrateDB v5.9.3

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 61                | 739       |

CrateDB v5.10.2

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 35                | 431       |

The average size of a shard is ~35MiB, `35 * 12 = ~420`.

> You can check the actual size of a shard locally; querying `select table_name, path from sys.shards`
> shows the file path of the shards.

Reviewer: ```suggestion
You can check the actual size of a shard by inspecting the filesystem.
Use `select table_name, path from sys.shards` to show the file path of the individual shards.
```

)
```

The index can only be disabled when the table is created, if the table already exists, and it can’t

> be deleted, it will have to be re-created.

Reviewer: Is it possible to delete an index afterwards? Doesn't the table _always_ need to be re-created?

Author: Ah, I see your point, will amend. `and it can’t
be deleted` goes out

> WARNING: Dropping the table deletes the data, make sure that the copy was done correctly.

> > INFO: Indexes can’t be re-added after table creation.

Reviewer: ```suggestion
:::{attention}
Indexes can’t be re-added after table creation.
:::
```

Author: Well its bad wording, they cannot be added after a table is created without them; @hlcianfagna wording is better.

### Effects on storage

CrateDB v5.9.3

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 53                | 639       |

CrateDB v5.10.2

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 39                | 471       |

Data was reduced `~13.11%` in v5.9 similar to `no_index`.
Data was increased `~10%` in v.5.10

> > NOTE: In v5.10, we don’t recommend that you turn off column storage.

Reviewer: ```suggestion
:::{note}
In v5.10, we don’t recommend that you turn off column storage.
:::
```

> ## Upgrade to v5.10 and Optimize table

Reviewer: Maybe?
```suggestion
## Upgrade to v5.10 and `OPTIMIZE TABLE`
```

> In v5.10 we introduced a new table storage format that can reduce storage up to 50% in comparison

Reviewer: ```suggestion
CrateDB v5.10 introduced a new table storage format that can reduce storage up to 50% in comparison
```

> In v5.10 we introduced a new table storage format that can reduce storage up to 50% in comparison
> to previous versions.
> 
> The easiest way to gain massive storage reduction is to update to the latest version available. 

Reviewer: I would suggest to remove the additional line break here.
```suggestion
In v5.10 we introduced a new table storage format that can reduce storage up to 50% in comparison
to previous versions. The easiest way to gain massive storage reduction is to update to the latest version available.
```

Author: You were right! It didn't change storage usage until I merged segments.

Additionally, as part of CrateDB's background operations, index segments are merged and optimized;
you would typically see storage usage being automatically reduced as more data gets added to the
table. Nonetheless, you can trigger these optimizations manually by running:

> `OPTIMIZE table <table_name> WITH (max_num_segments=1)`

Reviewer: Just a little more uppercasing.
```suggestion
`OPTIMIZE TABLE <table_name> WITH (max_num_segments=1)`
```

### Effects on storage

CrateDB v5.9.3

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 53                | 639       |

CrateDB v5.10.9

| avg_mib_per_shard | total_mib |
|-------------------|-----------|
| 39                | 471       |

Data was reduced `~13.11%` in v5.9 similar to `no_index`
Data was increased `~10%` in v5.10

:::{note}

> In v5.10, we don’t recommend that you turn off column storage.

Author: Yes, will do. I will mostly quote Alan:

Alan: _columnstore=off thing I would expect to not work any more, as if there's no columnstore then we need to store the data somewhere else, and columnstores will be more efficient in terms of disk usage as they are designed specifically for each data type.  Previously it was in the source and therefore duplicated, but now there is no source, so there's pretty much no advantage at all to disabling the column store._
Me: _That's why columnstore=off takes more space :)?_
Alan: _Yep. ..._

> In v5.10, we don’t recommend that you turn off column storage.

Reviewer: Thanks for explaining that!

Cheers, @romseygeek.

CrateDB's default settings are optimized for performance.

> If some columns are never used for aggregations or groupings, there will be no performance penalty.

Author: I will follow up this later.
